### ISSUE TITLE
Add optional function comment sniff for missing type-hint in @param

### ISSUE THREAD

**Issue by mindplay-dk** [squizlabs/PHP_CodeSniffer#1153]
Like the title says - this would optionally catch and warn about the following:

```php
/**
 * @param $bar
 */
function foo($bar) {}
```

The problem is with `@param $bar` which is missing a type-hint.

On our team, this is considered an error, because everyone uses PHP Storm, and it auto-generates doc-blocks like this one, and people forget to complete them.

I could PR and add this feature. Let me know?

**Comment by aik099** [comment]
The `Generic.Comments.DocComment` sniff should already do that and many other things.

**Comment by mindplay-dk** [comment]
> The `Generic.Comments.DocComment` sniff should already do that and many other things.

Doesn't seem to?

I poked through the code, and the only thing I can see pertaining to `@param` is something about the order of tag types.

**Comment by mindplay-dk** [comment]
Okay, it looks like you mean `Squiz.Commenting.FunctionComment`?

It has [a bunch of validations](https://github.com/squizlabs/PHP_CodeSniffer/blob/master/CodeSniffer/Standards/Squiz/Sniffs/Commenting/FunctionCommentSniff.php) for type-hints.

Unfortunately, these checks are much too condensed - some of these checks are covering a *ton* of different things, i.e. not enough granularity for this to be of any use to us.

For example, `Squiz.Commenting.FunctionComment.InvalidTypeHint` complains about whitespace and capitalization of the comment body, which is much too nit-picky and not what I need. It also demands `boolean` instead of `bool`, and `integer` instead of `int`, which seems plain wrong, given that the short form of those type-hints are the official "correct" ones used for static type-hinting in PHP 7.

Trying to reference specific rules doesn't seem to work for me at all.

For example, either this:

    <rule ref="Squiz.Commenting.FunctionComment.InvalidTypeHint"/>

Or this:

    <rule ref="Squiz.Commenting.FunctionComment.IncorrectTypeHint"/>

Produces the exact same result??

**Comment by mindplay-dk** [comment]
I don't understand why, but it seems I can't enable just `Squiz.Commenting.FunctionComment.TypeHintMissing`?

I do this:

    <rule ref="Squiz.Commenting.FunctionComment.TypeHintMissing"/>

But it enables every single inspection from `Squiz.Commenting.FunctionComment` regardless.

I finally had to enable `Squiz.Commenting.FunctionComment` and then suppress every inspection *except* `TypeHintMissing`:

    <rule ref="Squiz.Commenting.FunctionComment">
        <exclude name="Squiz.Commenting.FunctionComment.ParamCommentFullStop"/>
        <exclude name="Squiz.Commenting.FunctionComment.ParamCommentNotCapital"/>
        <exclude name="Squiz.Commenting.FunctionComment.SpacingAfterParamType"/>
        <exclude name="Squiz.Commenting.FunctionComment.SpacingAfterParamName"/>
        <exclude name="Squiz.Commenting.FunctionComment.ThrowsNotCapital"/>
        <exclude name="Squiz.Commenting.FunctionComment.ThrowsNoFullStop"/>
        <exclude name="Squiz.Commenting.FunctionComment.Missing"/>
        <exclude name="Squiz.Commenting.FunctionComment.MissingParamComment"/>
        <exclude name="Squiz.Commenting.FunctionComment.MissingParamName"/>
        <exclude name="Squiz.Commenting.FunctionComment.MissingParamTag"/>
        <exclude name="Squiz.Commenting.FunctionComment.MissingReturn"/>
        <exclude name="Squiz.Commenting.FunctionComment.IncorrectParamVarName"/>
        <exclude name="Squiz.Commenting.FunctionComment.InvalidReturn"/>
        <exclude name="Squiz.Commenting.FunctionComment.EmptyThrows"/>
    </rule>

That works, as far as enabling the correct inspection, but I'm concerned that any new inspections added in the future (not on that list) will now automatically be enabled.

Also, it seems to produce a lot of *false* inspections - for example, for [this file](https://gist.github.com/mindplay-dk/be76016ba4ff328521ef7761470a8585), I receive the following inspections:

```
<file name="C:\workspace\test\score\src\KodusClient.php">
 <error line="71" column="12" severity="error" message="Type hint &quot;array&quot; missing for $headers" source="Squiz.Commenting.FunctionComment.TypeHintMissing"/>
 <error line="83" column="12" severity="error" message="Type hint &quot;array&quot; missing for $headers" source="Squiz.Commenting.FunctionComment.TypeHintMissing"/>
 <error line="95" column="12" severity="error" message="Type hint &quot;array&quot; missing for $headers" source="Squiz.Commenting.FunctionComment.TypeHintMissing"/>
 <error line="107" column="12" severity="error" message="Type hint &quot;array&quot; missing for $headers" source="Squiz.Commenting.FunctionComment.TypeHintMissing"/>
</file>
```

None of those functions have any type-hints missing.

In fact, I haven't seen a valid inspection yet for a type-hint that was actually missing.

**Comment by gsherwood** [comment]
> I don't understand why, but it seems I can't enable just Squiz.Commenting.FunctionComment.TypeHintMissing?
> 
> I do this:
> 
> `<rule ref="Squiz.Commenting.FunctionComment.TypeHintMissing"/>`
>
> But it enables every single inspection from Squiz.Commenting.FunctionComment regardless.

That syntax is only available in the 3.0 version.

**Comment by mindplay-dk** [comment]
> That syntax is only available in the 3.0 version.

Ah, okay.

I've upgraded to `3.0.0RC1` now, and have it configured for just `TypeHintMissing`:

    <rule ref="Squiz.Commenting.FunctionComment.TypeHintMissing"/>

Unfortunately, nothing has changed with regards to the sample file and false inspections quoted above - `TypeHintMissing` is still producing false positives.

It appears this inspection is actually catching two different cases: one where the type-hint is missing from the php-doc block, and one where a static type-hint is missing from the actual parameter declaration in the code?

These are two very different error-conditions.

For various different reasons, it's not always possible to type-hint correctly with static type-hint - less so in PHP 7, where we can type-hint scalar arguments, but there are still examples where you cannot type-hint statically for some reason, such as having to use an upper or lower bound on an argument you inherited from a generic interface.

Am I correct that these two types of error are being reported as the same error?

To be clear: all I want is an inspection for `@param` doc-blocks with no type-hint, such as `@param $foo` - for some reason Php Storm auto-generates this junk, and our devs frequently forget to correct them... As for static type-hints on the other hand, as explained, we can't really enforce anything, as there would be literally hundreds of cases where devs would have to suppress these inspections with a tag...

**Comment by gsherwood** [comment]
> Unfortunately, nothing has changed with regards to the sample file and false inspections quoted above - TypeHintMissing is still producing false positives.

No, it wouldn't. I was only commenting on the config you were using. I haven't had time to look at this.

**Comment by mindplay-dk** [comment]
@gsherwood OK, well, the granularity issue was already addressed by 2.0, and the feature itself is actually there, so I will close this issue and reopen as a bug-report specifically.

thanks!
